src/controller/java/src/chip/devicecontroller/model/NodeState.java

```diff
@@ -44,14 +44,31 @@ private void addAttribute(
 
     ClusterState clusterState = endpointState.getClusterState(clusterId);
     if (clusterState == null) {
-      clusterState = new ClusterState(new HashMap<>());
+      clusterState = new ClusterState(new HashMap<>(), new HashMap<>());
       endpointState.getClusterStates().put(clusterId, clusterState);
     }
 
     // This will overwrite previous attributes.
     clusterState.getAttributeStates().put(attributeId, attributeStateToAdd);
   }
 
+  private void addEvent(int endpointId, long clusterId, long eventId, EventState eventStateToAdd) {
+    EndpointState endpointState = getEndpointState(endpointId);
+    if (endpointState == null) {
+      endpointState = new EndpointState(new HashMap<>());
+      getEndpointStates().put(endpointId, endpointState);
+    }
+
+    ClusterState clusterState = endpointState.getClusterState(clusterId);
+    if (clusterState == null) {
+      clusterState = new ClusterState(new HashMap<>(), new HashMap<>());
+      endpointState.getClusterStates().put(clusterId, clusterState);
+    }
+
+    // This will overwrite previous events.
+    clusterState.getEventStates().put(eventId, eventStateToAdd);
```

Comment:
events should be stored in array. in real user scenario, you may receive a burst of events, for example, lock/unlock, we would like to record all events.